torch/_torch_docs.py

```diff
@@ -3385,6 +3385,55 @@ def merge_dicts(*dicts):
     array([-1,  2,  3])
 """)
 
+add_docstr(torch.from_buffer,
+           r"""
+from_buffer(buffer, dtype=None, count=-1, offset=0, *, device=None, requires_grad=False) -> Tensor
+
+Create a 1-dimensional :class:`Tensor` from an object that implements
+the Python buffer protocol.
+
+Interpret :attr:`buffer` as a 1-dimensional buffer of type :attr:`dtype`
+with :attr:`count` elements after the first :attr:`offset` bytes.
+
+The returned tensor and buffer share the same memory. Modifications to
+the tensor will be reflected in the buffer and vice versa. The returned
+tensor is not resizable.
+
+.. note::
+    This function increments the reference count for the object that
+    owns the shared memory. Therefore, such memory will not be deallocated
+    before the returned tensor goes out of scope.
+
+.. warning::
+    Passing a pointer that lives in a different device than the specified
```

Comment:
This seems bad. Is there some way we can address this? Alternatively, should we restrict this to CPU only for now?

Reply:
Actually, I guess `asarray` might solve this with its copy kwarg? Should we be assuming that objects implementing the Python buffer protocol are on the CPU, and make a copy if `device` is a CUDA device?

cc @rgommers for this question, too

Reply:
I'm not sure this warning makes sense to me. There is no raw pointer passed anywhere by the user. Maybe what is meant is that if `buffer` and `device` are mismatching, things go wrong? Better to say that then.

There's this fundamental issue here that the doc says _"a Python object that exposes the buffer interface"_, but the buffer protocol knows nothing about devices. There's only a couple of libraries that use the buffer protocol on non-CPU devices, like `mpi4py` - here's a random example I could find: https://gist.github.com/lebedov/8514d3456a94a6c73e6d.

> Actually, I guess `asarray` might solve this with its copy kwarg? Should we be assuming that objects implementing the Python buffer protocol are on the CPU, and make a copy if `device` is a CUDA device?

Never make a copy I'd say - this is a zero-copy thing. You can do it on GPU if `device` is handled correctly, it's just unfortunate that `device` has to be given by the end user which is a chance to make a mistake with the consequence of grabbing the wrong memory. That's in contrast to DLPack, where it's the library implementer's responsibility to handle devices.

Either way, let's never make a copy - just share memory or raise an exception. That's also consistent with `__array__`, `__cuda_array_interface__`, etc.

Reply:
We're trying to replace existing functionality here though, right? One of the `Tensor.__init__` usages? How did that handle this issue?

Reply:
> Maybe what is meant is that if buffer and device are mismatching, things go wrong? Better to say that then.

Yes, that's indeed the point. The way you put it is better. I will change the doc.

> There's this fundamental issue here that the doc says "a Python object that exposes the buffer interface", but the buffer protocol knows nothing about devices.

It's as you describe next. Basically, it's the user responsibility here to get the device right (which is unfortunate).

> Either way, let's never make a copy - just share memory or raise an exception.

Agreed. I also think we shouldn't copy, here (there's the tensor factory for that).

> We're trying to replace existing functionality here though, right? One of the Tensor.__init__ usages? How did that handle this issue?

The idea was to replace `new(Storage)` (construct a tensor, given a `Storage`). In _torchvision_, it was used in conjunction with `Storage.from_buffer`. This implementation:

- Assumed the buffer lived on CPU
- Copied the data to this new storage

https://github.com/pytorch/pytorch/blob/010bcb4c2dbd0bc550252ea9cac942006758fb5f/torch/csrc/generic/StorageMethods.cpp#L97